field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -216,30 +220,25 @@ checkExpr m = do
     let uniMismatch = typeMismatch tM uni mUni
     withGeqUniM uni mUni uniMismatch $ tM
 
+checkProgram
+    :: forall m f. (MonadTypeChecker m, TextField f)
+    => R.Program R.Var f -> m (T.Program f)
+checkProgram =
+    liftA T.Program . liftA T.Statements . mapM checkStatement . R.unStatements .  R.unProgram
```

Comment:
`fmap (T.Program . T.Statements)` instead of `liftA T.Program . liftA T.Statements`?

Reply:
What if we define

```
newtype Statements stmt = Statements [stmt]

newtype Program stmt = Program (Statements stmt)
```

? Then we can reuse `Statements` and `Program` for both the typed and raw ASTs and this definition will be just `traverse checkStatement`.

The downside is that if the ASTs get out of sync, then this won't work anymore. But I'm not sure how likely that is.

Reply:
Hmm, but in which module should I put those?  `Field/Core.hs`?

Reply:
Hmm, should we also parameterise those newtype by the field?

Therefore yielding something akin to
```haskell
newtype Statements stmt f = Statements [stmt f]
```

Reply:
> Hmm, but in which module should I put those? `Field/Core.hs`?

Whatever you think is appropriate. Might be `Field.Raw.Core` as well, since the typed language depends on the raw one anyway.

> Hmm, should we also parameterise those newtype by the field?

I think it shouldn't be required? You just instantiate `stmt` with a particular `Statement` (either raw or typed one) and it's what contains the `f`.